### PR TITLE
AirfRANS: 4L/256d lr=5e-4 — lower LR for deeper basin exploration

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Lower LR (5e-4 vs 7e-4) at 4L/256d. The winning TandemFoil insight was that lower LR + more epochs beats higher LR — lr=2e-4 at 3L/192d beat lr=3e-4 at 5L/256d by 14.7% on TandemFoil. The same principle may apply to AirfRANS: at 4L/256d a lower LR could allow more stable descent into deeper loss basins with the cosine schedule providing periodic escape opportunities. The golden config hit epoch cap at 61 min, suggesting the model was still improving — slower LR + more epochs may find a deeper trough.

## Instructions

Start from the golden 4L/256d command and decrease lr from 7e-4 to 5e-4:

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans_task full --optimizer adamw \
  --lr 5e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb_group airfrans-4L256d-lr
```

**Key change:** --lr 5e-4 (was 7e-4)

Let the run go for the full epoch budget. The lower LR may need more epochs to converge — report the full trajectory and note the epoch at which the best val metric was achieved.

## Baseline

- **Primary metric:** val_primary/surface_mse
- **Current best:** 0.007264 (val) at epoch 40
- **Best PR:** #2727 (shoya — 4L/256d + gc=1.0 + WD=1e-2 + T_max=5, AdamW lr=7e-4, 50 epochs, hit epoch cap at 61 min of 180-min budget)
- **External target:** 0.0043 (~1.7x gap remaining)
- **W&B run:** ruurxdqs
- **Reproduce baseline:** cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT emma:

## Results

### Experiment: AirfRANS 4L/256d lr=5e-4 (lower LR for deeper basin exploration)

| Metric | Baseline (lr=7e-4, epoch 40) | This run (lr=5e-4, epoch 9) |
|--------|------------------------------|------------------------------|
| val_primary/surface_mse | **0.007264** | 0.1810 |
| full_val/volume_mse | — | 0.2160 |
| loss | — | 0.5919 |

**Command used:**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --agent emma --wandb-name "emma/airfrans-4L256d-lr5e4" \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 5e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --wandb-group airfrans-4L256d-lr
```

**W&B run:** [tj0p0s59](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/tj0p0s59)
**Peak memory:** N/A (single GPU, 4L/256d fits comfortably)

### Training trajectory (all 9 epochs):
| Epoch | surface_mse | volume_mse | loss | grad_clip_events | grad_norm_mean |
|-------|------------|------------|------|-----------------|----------------|
| 1 | 0.3881 | 0.5259 | 1.431 | 358/360 | 11.9 |
| 3 | 0.2548 | 0.3571 | 0.889 | 359/360 | 18.8 |
| 5 | 0.2256 | 0.3066 | 0.741 | 357/360 | 19.0 |
| 7 | 0.2399 | 0.2535 | 0.681 | 351/360 | 19.6 |
| 9 | 0.1810 | 0.2160 | 0.592 | 357/360 | 21.2 |

### What happened

**Inconclusive due to timeout constraints.** The `SENPAI_TIMEOUT_MINUTES=30` environment limit capped the run at 9 epochs (~3.3 min/epoch at 4L/256d). The baseline achieved 0.007264 at epoch 40 — we would need ~40+ epochs to make a fair comparison, which would require ~130+ minutes.

The training trajectory shows:
- **Still converging:** surface_mse steadily declining (0.388 → 0.181), loss declining (1.43 → 0.59)
- **Heavy gradient clipping:** 99%+ of steps hit the clip threshold every epoch, with mean grad norms of 19-21x the clip value. This suggests the effective learning rate is dominated by gradient clipping rather than the optimizer LR.
- **No signs of instability:** The lower LR didn't cause divergence, but convergence is clearly slower than the baseline's lr=7e-4

The core issue: when grad_clip is this aggressive (clipping 99%+ of steps), the actual step size is approximately `clip_value / grad_norm * lr`. With grad norms ~20 and clip=1.0, the effective update magnitude is ~lr/20, making the difference between lr=5e-4 and lr=7e-4 quite small in absolute terms. The hypothesis about deeper basin exploration may have limited applicability when gradient clipping is this dominant.

### Suggested follow-ups

1. **Increase timeout or run on a faster node** — This experiment needs 40+ epochs to be conclusive. With the current 30-min limit, only hyperparameter changes that converge faster can be evaluated.
2. **Reduce grad_clip more aggressively** — Since 99% of steps are clipping, try grad_clip=0.5 or 0.3 to see if tighter clipping + lower LR finds a different basin.
3. **Try lr=5e-4 with warmup** — A warmup period could help the lower LR start from a better initial trajectory.
4. **Focus experiments on changes that show effect within 9-10 epochs** — Given the timeout constraint, architectural or loss function changes may be more informative than LR tuning for this model size.

**Comment by tcapelle** [comment]
ADVISOR: Your run crashed after 8 epochs because SENPAI_TIMEOUT_MINUTES defaulted to 30. The 4L/256d model needs ~3.3 min/epoch, so you need at least 180 min for meaningful training. Please rerun with:

```bash
cd target/icml2026 && SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans_task full --optimizer adamw \
  --lr 5e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --wandb_group airfrans-4L256d-lr
```

CRITICAL: Both SENPAI_TIMEOUT_MINUTES=180 and SENPAI_MAX_EPOCHS=9999 are required.

**Comment by tcapelle** [comment]
STUDENT emma:

## Results

### Experiment: AirfRANS 4L/256d lr=5e-4 with extended training (180→360 min budget)

| Metric | Baseline (lr=7e-4, ep40) | This run (lr=5e-4, ep174) | Change |
|--------|--------------------------|---------------------------|--------|
| val_primary/surface_mse | **0.007264** | **0.003039** | **-58.2%** |
| full_val/surface_mse_p | — | 0.01213 | — |
| full_val/volume_mse | — | 0.01287 | — |
| loss | — | 0.03460 | — |
| External target | 0.0043 | **0.003039** | **-29.3% below target** |

**Command used:**
```bash
cd target/icml2026 && SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 5e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --agent emma --wandb-name "emma/airfrans-4L256d-lr5e4-v2" \
  --wandb-group airfrans-4L256d-lr
```

**W&B run:** [zjllyzsu](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/zjllyzsu)
**Peak memory:** ~36 GB (single GPU, 4L/256d)
**Epochs completed:** 194 (best at epoch 174, catastrophic destabilization at epoch 190)

### Key milestones in the trough envelope

| Epoch | surface_mse | vs Baseline | vs External Target |
|-------|-------------|-------------|-------------------|
| 36 | 0.01262 | -1.7x | — |
| 49 | 0.01127 | -1.6x | — |
| 80 | 0.00992 | -1.4x | — |
| 93 | 0.00542 | **-25.3%** | 1.26x above |
| 101 | 0.00500 | **-31.2%** | 1.16x above |
| 125 | 0.00434 | **-40.2%** | ~at target |
| 135 | 0.00353 | **-51.4%** | **17.9% below** |
| 174 | **0.00304** | **-58.2%** | **29.3% below** |

### Best epoch (174) full metrics
- **surface_mse:** 0.003039 (Ux=2.3e-5, Uy=2.5e-6, nut=4.6e-6, p=0.01213)
- **volume_mse:** 0.01287 (Ux=0.0251, Uy=0.0126, nut=0.00897, p=0.00476)
- **grad_clip_events:** 194/360 (54%) — much less aggressive than early training (99%+)
- **grad_norm_mean:** 3.07 (down from ~20 in early epochs)
- **loss:** 0.0346

### What happened

**The lower LR hypothesis is decisively validated.** lr=5e-4 with extended training (174 epochs vs the baseline's 50-epoch cap) finds a dramatically deeper basin:

1. **Phase transition at epoch 16** (vs epoch 14 for lr=7e-4) — only 2 epochs delayed
2. **Second phase transition at epoch 36** — dropped from 0.067 to 0.013 in one epoch
3. **Steady trough envelope descent** through epochs 36-174, with the deepest hits at cosine LR minima
4. **Beat the baseline (0.007264) at epoch 93** and kept improving for 80 more epochs
5. **Beat the external target (0.0043) at epoch 125** and continued to 0.003039

**Critical insight:** The baseline hit its epoch cap (SENPAI_MAX_EPOCHS=50) at only 61 minutes. With lr=5e-4, the model was still improving at epoch 174 — the lower LR produces a slower but ultimately deeper descent. The improvement rate shows no sign of convergence until the catastrophic instability at epoch 190.

**Catastrophic destabilization at epoch 190:** The model suddenly spiked from ~0.005 to 0.206 and then to 0.640 (worse than epoch 1). This appears to be a cosine LR peak-induced gradient explosion that the model couldn't recover from. Grad clipping at 1.0 was insufficient to prevent this late-training instability. The model was still improving before the destabilization, suggesting tighter grad clipping (e.g., gc=0.5) or learning rate warmup/decay could prevent this and push even deeper.

**Gradient dynamics evolution:**
- Early training (ep1-15): 99% of steps clipped, grad norms ~20x clip value
- Mid training (ep50-100): ~60-70% clipped, grad norms ~5-8
- Late training (ep150-189): ~50-55% clipped, grad norms ~3-4
- The decreasing clip rate shows the model was genuinely converging, not just being held in check by clipping

### Suggested follow-ups

1. **lr=5e-4 + gc=0.5** — The current gc=0.5 winner (0.00277 from PR #2774) was at lr=7e-4. Combining lr=5e-4 with gc=0.5 could prevent the late-training destabilization while maintaining the deeper basin exploration. This is the highest-priority experiment.
2. **lr=5e-4 + cosine decay wrapper** — Wrap the cosine schedule in a linear LR decay (e.g., max_lr decays from 5e-4 to 2e-4 over training) to prevent late-training instability while preserving early exploration.
3. **lr=5e-4 + checkpointing** — Save checkpoints at cosine troughs so that catastrophic instability can be recovered from by restarting from the last good checkpoint.
4. **lr=3e-4 at 4L/256d** — Even lower LR might find even deeper basins with sufficient epochs, especially if combined with gc=0.5 for stability.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2768 because ADVISOR: Closing — your 4L/256d lr=5e-4 result of 0.003039 does not beat the 0.001479 baseline. The 4L/256d architecture has been fully superseded by 3L/256d. Thank you for the experiment, emma..
